### PR TITLE
[AD]: Created the logs-scheduler and plugged it to autodiscovery

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

* Created a log-scheduler to handle configs provided by autodiscovery.
* Updated the file config provider to scope logs-config in "logs" attribute
* Removed the old logs logic to retrieve config from files
* Refactored `pkglogs/config`
* Updated all inputs to consume a stream of log sources
### Motivation

Use only one solution to configure the agent and benefits from the existing logics instead of reimplementing all of them.

### Additional Notes

* Only support config from files at the moment.

### CODE REVIEW

File: pkg/logs/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	var configs []*config.LogsConfig
+	var err error
+
+	switch integrationConfig.Provider {
+	case providers.File:
+		configs, err = config.ParseYaml(integrationConfig.LogsConfig)
+	case providers.Docker, providers.Kubernetes:
+		configs, err = config.ParseJSON(integrationConfig.LogsConfig)
+	default:
+		return nil, fmt.Errorf("parsing logs-config from %v is not supported yet.", integrationConfig.Provider)
```
Comment (mnshdw) [inline pkg/logs/scheduler.go]: nit: Can we simplify to err = fmt.Errorf(...) ?

File: pkg/logs/input/docker/container.go [hunk 2]
```diff
@@ -31,8 +31,17 @@ func NewContainer(container types.Container) *Container {
...
-		cfg, err := config.Parse(label)
-		if err != nil {
+		configs, err := config.ParseJSON([]byte(label))
+		if err != nil || len(configs) == 0 {
+			log.Errorf("Invalid docker label for container %v: %v", c.Container.ID, err)
+			return nil
+		}
+		cfg := configs[0]
+		if isValid, err := config.Validate(cfg); !isValid {
+			log.Errorf("Invalid docker label for container %v: %v", c.Container.ID, err)
```
Comment (mnshdw) [inline pkg/logs/input/docker/container.go]: I think we could afford a different error message.

**Review by mnshdw — COMMENTED** [review]
Looking very good at the moment, I only left a few nits.

File: pkg/logs/config/config.go [hunk 3]
```diff
@@ -44,6 +26,7 @@ func buildLogSources(ddconfdPath string, collectAllLogsFromContainers bool, tcpF
 		sources = append(sources, containersSource)
```
Comment (tmichelet) [inline pkg/logs/config/config.go]: Could it be a problem that this source would be the first in the list?

File: cmd/agent/common/autoconfig.go [hunk 4]
```diff
@@ -38,6 +39,11 @@ func SetupAutoConfig(confdPath string) {
 	// registering the check scheduler
 	metaScheduler.Register("check", collector.InitCheckScheduler(Coll))
 
+	// registering the logs scheduler
+	if scheduler, err := logs.GetScheduler(); err == nil {
```
Comment (mfpierre) [inline cmd/agent/common/autoconfig.go]: should we log `err`?

File: pkg/logs/config/integration_config.go [hunk 5]
```diff
@@ -79,163 +65,44 @@ type LogsConfig struct {
...
+	return true, nil
 }
 
-// compileProcessingRules compiles all processing rules regular expression
-func compileProcessingRules(rules []LogsProcessingRule) error {
+// Compile compiles all processing rules regular expression.
+func Compile(config *LogsConfig) error {
+	rules := config.ProcessingRules
 	for i, rule := range rules {
-		if rule.Pattern == "" {
```
Comment (tmichelet) [inline pkg/logs/config/integration_config.go]: why are we deleting this line?

File: pkg/logs/scheduler.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+// NewScheduler returns a new scheduler.
+func NewScheduler(sources *config.LogSources) *Scheduler {
+	return &Scheduler{
+		sources: sources,
+	}
+}
+
+// Start does nothing.
+func (s *Scheduler) Start() {}
```
Comment (mfpierre) [inline pkg/logs/scheduler.go]: this was removed from this interface, you can drop it

File: pkg/logs/config/integration_config_test.go [hunk 7]
```diff
@@ -8,161 +8,67 @@
...
-	assert.Equal(t, "include_at_match", iRule.Type)
-	assert.Equal(t, "include_datadoghq", iRule.Name)
-	assert.Equal(t, ".*@datadoghq.com$", iRule.Pattern)
-	re = iRule.Reg
-	assert.True(t, re.MatchString("bob@datadoghq.com"))
-	assert.False(t, re.MatchString("bob"))
+func TestValidateShouldFailWithInvalidConfigs(t *testing.T) {
+	invalidConfigs := []*LogsConfig{
+		{Type: FileType},
+		{Type: TCPType},
```
Comment (tmichelet) [inline pkg/logs/config/integration_config_test.go]: that seems way better!

File: pkg/autodiscovery/providers/file.go [hunk 8]
```diff
@@ -288,7 +288,9 @@ func GetIntegrationConfigFromFile(name, fpath string) (integration.Config, error
 
 	// If logs was found, add it to the config
 	if cf.LogsConfig != nil {
-		rawLogsConfig, _ := yaml.Marshal(cf.LogsConfig)
+		logsConfig := make(map[string]interface{})
+		logsConfig["logs"] = cf.LogsConfig
+		rawLogsConfig, _ := yaml.Marshal(logsConfig)
```
Comment (mfpierre) [inline pkg/autodiscovery/providers/file.go]: assign directly `config.LogsConfig` and drop the line below?

File: pkg/autodiscovery/providers/file.go [hunk 9]
```diff
@@ -288,7 +288,9 @@ func GetIntegrationConfigFromFile(name, fpath string) (integration.Config, error
 
 	// If logs was found, add it to the config
 	if cf.LogsConfig != nil {
-		rawLogsConfig, _ := yaml.Marshal(cf.LogsConfig)
+		logsConfig := make(map[string]interface{})
+		logsConfig["logs"] = cf.LogsConfig
```
Comment (tmichelet) [inline pkg/autodiscovery/providers/file.go]: could we avoid hardcoding `logs`? we have it defined as a constant later

**Review by mfpierre — COMMENTED** [review]
Left some comments, nice to see this shaping up 😎

**Review by tmichelet — COMMENTED** [review]
This is going in the right direction.

I am wondering if we need to keep all the scanners around, or if we could rely on the scheduler to spawn new tailers. It can definitely be done in a later step though.

We need to be extra careful around docker, as it will be easy to break the current behavior. Before we merge let's test all scenarii thoroughly

File: pkg/logs/config/integration_config.go (hunk 5 above)
Comment (ajacquemot) [reply pkg/logs/config/integration_config.go]: It's validation logic, not compilation

File: pkg/logs/config/config.go (hunk 3 above)
Comment (ajacquemot) [reply pkg/logs/config/config.go]: It's not anymore, I implemented a logic to assign the right source to the right container, see https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/docker/container.go#L63-L76

File: pkg/logs/config/integration_config.go (hunk 5 above)
Comment (tmichelet) [reply pkg/logs/config/integration_config.go]: ah I see we moved it

File: pkg/logs/config/config.go (hunk 3 above)
Comment (tmichelet) [reply pkg/logs/config/config.go]: 👍

**Comment by ajacquemot** [comment]
I agree with you @tmichelet regarding removing the scanner logic, I am working on it right now on another branch, the main goal of this PR is to have the logs-agent working with the AD and the file provider.

**Review by tmichelet — DISMISSED** [review]

File: cmd/agent/common/autoconfig.go [hunk 10]
```diff
@@ -37,6 +38,11 @@ func SetupAutoConfig(confdPath string) {
 	// registering the check scheduler
 	metaScheduler.Register("check", collector.InitCheckScheduler(Coll))
 
+	// registering the logs scheduler
+	if logs.IsAgentRunning() {
+		metaScheduler.Register("logs", logs.GetScheduler())
```
Comment (mfpierre) [inline cmd/agent/common/autoconfig.go]: ✨

**Review by mfpierre — DISMISSED** [review]

**Review by arbll — DISMISSED** [review]

**Pull request merged**